core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -0,0 +1,437 @@
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.internal.ArmeriaHttpUtil;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.HttpStatusException;
+import com.linecorp.armeria.server.RequestTimeoutException;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerListenerAdapter;
+import com.linecorp.armeria.server.ServiceConfig;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.TransientService;
+
+import io.netty.util.AsciiString;
+import io.netty.util.concurrent.ScheduledFuture;
+
+/**
+ * An {@link HttpService} that responds with HTTP status {@code "200 OK"} if the server is healthy and can
+ * accept requests and HTTP status {@code "503 Service Not Available"} if the server is unhealthy and cannot
+ * accept requests. The default behavior is to respond healthy after the server is started and unhealthy
+ * after it started to stop.
+ *
+ * <h3>Long-polling support</h3>
+ *
+ * <p>A client that sends health check requests to this service can send a long-polling request to get notified
+ * immediately when a {@link Server} becomes healthy or unhealthy, rather than sending health check requests
+ * periodically.</p>
+ *
+ * <p>To wait until a {@link Server} becomes unhealthy, i.e. wait for the failure, send an HTTP request with
+ * two additional headers:
+ * <ul>
+ *   <li>{@code If-None-Match: "healthy"}</li>
+ *   <li>{@code Prefer: wait=<seconds>}
+ *     <ul>
+ *       <li>e.g. {@code Prefer: wait=60}</li>
+ *     </ul>
+ *   </li>
+ * </ul></p>
+ *
+ * <p>To wait until a {@link Server} becomes healthy, i.e. wait for the recovery, send an HTTP request with
+ * two additional headers:
+ * <ul>
+ *   <li>{@code If-None-Match: "unhealthy"}</li>
+ *   <li>{@code Prefer: wait=<seconds>}</li>
+ * </ul></p>
+ *
+ * <p>The {@link Server} will wait up to the amount of seconds specified in the {@code "Prefer"} header
+ * and respond with {@code "200 OK"}, {@code "503 Service Unavailable"} or {@code "304 Not Modified"}.
+ * {@code "304 Not Modifies"} signifies that the healthiness of the {@link Server} did not change.
+ * Once the response is received, the client is supposed to send a new long-polling request to continue
+ * watching the healthiness of the {@link Server}.</p>
+ *
+ * <p>All health check responses will contain a {@code "armeria-lphc"} header whose value is the maximum
+ * allowed value of the {@code "Prefer: wait=<seconds>"} header. {@code 0} means long polling has been
+ * disabled. {@code "lphc"} stands for long-polling health check.</p>
+ *
+ * @see HealthCheckServiceBuilder
+ */
+public final class HealthCheckService implements HttpService, TransientService<HttpRequest, HttpResponse> {
+
+    private static final Logger logger = LoggerFactory.getLogger(HealthCheckService.class);
+    private static final AsciiString ARMERIA_LPHC = HttpHeaderNames.of("armeria-lphc");
```

Comment:
I can't think of anything better other than `armeria-long-polling`, which is shorter than the previous one but lacks health check aspect